File: C++/0078.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    std::vector<std::vector<int>> merge(std::vector<std::vector<int>>& intervals) {
        std::sort(intervals.begin(), intervals.end());
        int idx = 0;
        while(idx < intervals.size()-1){
            if(intervals[idx][1] >= intervals[idx+1][0]){
                intervals[idx] = {std::min(intervals[idx][0], intervals[idx+1][0]), std::max(intervals[idx][1], intervals[idx+1][1])};
                intervals.erase(intervals.begin() + idx + 1);
            } else {
                ++idx;
            }
        }
        return intervals;
    }
};

// Topic: Array, Sorting
// Time Complexity: O(nlogn)
// Space Complexity: O(1)

File: Python/0078.py
from typing import List
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        idx = 0
        while idx < len(intervals)-1:
            if intervals[idx][1] >= intervals[idx+1][0]:
                intervals[idx] = [min(intervals[idx][0], intervals[idx+1][0]), max(intervals[idx][1], intervals[idx+1][1])]
                del intervals[idx+1]
            else:
                idx += 1

        return intervals

# Topic: Array, Sort
# Time: O(nlogn)
# Space: O(1)
        